### PR TITLE
fix: update component types define, add children & add contextType, fix ts check error with @types/react v18

### PR DESCRIPTION
**Pull request by pointhalo**
…heck error with @types/react 18, #755

[中文模板 / Chinese Template](https://github.com/DouyinFE/semi-design/blob/main/.github/PULL_REQUEST_TEMPLATE.zh-CN.md)

- [x] I have read and followed [Pull Request Guidelines](https://github.com/DouyinFE/semi-design/blob/main/CONTRIBUTING-en-US.md#pull-request-guidelines) of the contributing guide.

### What kind of change does this PR introduce? (check at least one)

 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update
 - [ ] Refactor
 - [ ] Test Case
 - [ ] TypeScript definition update
 - [ ] Document improve
 - [ ] CI/CD improve
 - [ ] Branch sync
 - [ ] Other, please describe:

#### Reason for change （改动原因）

@types/react 18 make some breaking change in this PR (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210),
Major changes
- Removed React.SFC, use FC instead
- Removed StatelessComponent and used FunctionComponent uniformly
- **this.context changed from any to unknown**
- **React.Component and FunctionComponent remove the declaration of children**

The above changes will cause all users of types/react 18 to encounter type errors when using Semi

#### The point of change（改动重点）

- Append declaration to the interface of components that do not explicitly declare `children?: React.ReactNode | undefined`
- For components that do not explicitly declare a context, explicitly declare the type of the context

image

### Changelog
🇨🇳 Chinese
- Fix:  修复搭配 @types/react 18使用时 ，ts类型检查报错的问题, #755

---

🇺🇸 English
- Fix: Fix the problem of ts type checking error when used with @types/react 18, #755

### Checklist
- [ ] Test or no need
- [ ] Document or no need
- [ ] Changelog or no need

### Other
- [ ] Skip Changelog

### CODE REVIEW

**Comment by pointhalo** [comment]
image

**Comment by richtera** [comment]
These typedef changes seem also to affect a bunch of seemingly unrelated things. Is it no longer valid to pass more than one child to `<h5>` or `<div>`, for example? I do see a lot of mentions about PropsWithChildren but I didn't see any mention about normal dom elements being affect by this or did I miss this somewhere in the other defects?

**Comment by richtera** [comment]
These typedef changes seem also to affect a bunch of seemingly unrelated things. Is it no longer valid to pass more than one child to `<h5>` or `<div>`, for example? I do see a lot of mentions about PropsWithChildren but I didn't see any mention about normal dom elements being affect by this or did I miss this somewhere in the other defects?

File: packages/semi-ui/navigation/SubNavTransition.tsx [hunk 1]
```diff
@@ -13,7 +13,7 @@ const formatStyle = function formatStyle({ maxHeight, opacity }: { maxHeight: nu
 };
 
 export interface SubNavTransitionProps {
-    children?: React.ReactNode;
+    children?: React.ReactNode | ((transitionArgus?: any) => React.ReactNode);
```
Comment (DaiQiangReal) [inline packages/semi-ui/navigation/SubNavTransition.tsx:16]: spell

File: packages/semi-ui/navigation/SubNav.tsx [hunk 2]
```diff
@@ -162,9 +163,9 @@ export default class SubNav extends BaseComponent<SubNavProps, SubNavState> {
             notifyGlobalOpenChange: (...args) => this._invokeContextFunc('onOpenChange', ...args),
             notifyGlobalOnSelect: (...args) => this._invokeContextFunc('onSelect', ...args),
             notifyGlobalOnClick: (...args) => this._invokeContextFunc('onClick', ...args),
-            getIsSelected: itemKey => Boolean(!isNullOrUndefined(itemKey) && get(this.context, 'selectedKeys', []).includes(itemKey)),
+            getIsSelected: itemKey => Boolean(!isNullOrUndefined(itemKey) && get(this.context, 'selectedKeys', []).includes(itemKey as string)),
```
Comment (DaiQiangReal) [inline packages/semi-ui/navigation/SubNav.tsx:166]: String(itemKey)

File: packages/semi-ui/notification/NoticeTransition.tsx [hunk 3]
```diff
@@ -9,7 +9,7 @@ export type ArgsType<T> = T extends (...args: infer A) => any ? A : never;
 export interface NoticeTransitionProps{
     position?: NoticePosition;
     motion?: Motion<NoticeTransitionProps>;
-    children?: JSXElementConstructor<any> | React.ReactChildren;
+    children?: React.ReactNode | ((ChildArgus: any) => React.ReactNode);
```
Comment (DaiQiangReal) [inline packages/semi-ui/notification/NoticeTransition.tsx:12]: spell

File: packages/semi-animation-react/src/StyledTransition.tsx [hunk 4]
```diff
@@ -1,14 +1,20 @@
...
+    animateStyle?: StyledAnimateStyleType;
+    animateEvents?: (eventProps?: any) => void;
+}
+
 export interface StyledTransitionProps extends StyledAnimationProps {
     state?: string | boolean;
     enter?: string;
     leave?: string;
-    children?: React.ReactNode;
+    children?: React.ReactNode | ((ChildArgus: ChildFCType) => React.ReactNode | any);
```
Comment (DaiQiangReal) [inline packages/semi-animation-react/src/StyledTransition.tsx:17]: spell

File: packages/semi-animation-react/src/StyledTransition.tsx [hunk 5]
```diff
@@ -19,8 +25,8 @@ export interface StyledTransitionProps extends StyledAnimationProps {
 
 export interface StyledTransitionState {
     state: string | boolean;
-    lastChildren: React.ReactNode;
-    currentChildren: React.ReactNode;
+    lastChildren:  React.ReactNode | ((ChildArgus: ChildFCType) => React.ReactNode | any);
+    currentChildren: React.ReactNode | ((ChildArgus: ChildFCType) => React.ReactNode | any);
```
Comment (DaiQiangReal) [inline packages/semi-animation-react/src/StyledTransition.tsx:29]: spell

File: packages/semi-animation-react/src/Transition.tsx [hunk 6]
```diff
@@ -5,7 +5,7 @@ import React, { Component, isValidElement } from 'react';
 import noop from './utils/noop';
 
 export interface TransitionProps extends AnimationProps {
-    children?: React.ReactNode;
+    children?: React.ReactNode | ((ChildArgus: any) => any);
```
Comment (DaiQiangReal) [inline packages/semi-animation-react/src/Transition.tsx:8]: spell

File: packages/semi-ui/breadcrumb/item.tsx [hunk 7]
```diff
@@ -129,7 +137,7 @@ export default class BreadcrumbItem extends BaseComponent<BreadcrumbItemProps, B
         const showTooltip = this.getTooltipOpt();
         const icon = this.renderIcon();
         if (Boolean(children) && typeof children === 'string') {
-            const { opts, ellipsisPos, width } = showTooltip;
+            const { opts, ellipsisPos, width } = showTooltip as GetTooltipOptType;
```
Comment (DaiQiangReal) [inline packages/semi-ui/breadcrumb/item.tsx:140]: 是否可以标在 函数返回值上

Comment (pointhalo) [reply packages/semi-ui/breadcrumb/item.tsx:140]: 如果在 getTooltipOpt 上声明返回类型的话，也需要处理一下类型强转的问题（看起来是typeof后的类型没推断出来）。这里为了方便起见，就直接as了

image
image

File: packages/semi-animation-react/src/Transition.tsx (hunk 6 above)
Comment (pointhalo) [reply packages/semi-animation-react/src/Transition.tsx:8]: 已改

File: packages/semi-animation-react/src/StyledTransition.tsx (hunk 5 above)
Comment (pointhalo) [reply packages/semi-animation-react/src/StyledTransition.tsx:29]: 已改

**Review by DaiQiangReal — APPROVED** [review]

**Comment by pointhalo** [comment]
release v2.8.2

**Pull request merged**